Fixed traceback when not run as root

bradmw opened the pull request:
Looks like the way cli.py was automerged in #558 caused a traceback to be shown when Let's Encrypt is not run as root and the default configuration directories are used. This PR fixes that.

kuba commented:
As discussed in IRC this might require a bit more thinking due to cyclic dependency problems... :(

bradmw commented:
The most recent commit reimplements top level exception handling. In addition to cleaning the code up a bit, it does a better job of handling exceptions before arg parsing and logger setup. If an exception is thrown before logger setup and the exception is a custom exception type, only the contents of the error message are printed. If this exception is not a custom error type, the full exception and traceback is printed to stderr.

kuba commented on letsencrypt/cli.py:
```diff
@@ -642,8 +644,45 @@ def _setup_logging(args):
+        "Exiting abnormally:\n%s",
+        "".join(traceback.format_exception(exc_type, exc_value, trace)))
+
+    if issubclass(exc_type, errors.Error) and (not args or not args.debug):
```
please use `args is not None` rather than `not args`

kuba commented on letsencrypt/cli.py:
```diff
@@ -642,8 +644,45 @@ def _setup_logging(args):
+
+    if issubclass(exc_type, errors.Error) and (not args or not args.debug):
+        sys.exit(exc_value)
+    elif issubclass(exc_type, Exception) and args and not args.debug:
```
same as above

maybe could you also add docstring explaining behaviour under `args is None` and `args is not None`?

kuba commented on letsencrypt/cli.py:
```diff
@@ -642,8 +644,45 @@ def _setup_logging(args):
+        "".join(traceback.format_exception(exc_type, exc_value, trace)))
+
+    if issubclass(exc_type, errors.Error) and (not args or not args.debug):
+        sys.exit(exc_value)
```
how about trying to dump a file with a full traceback to `./letsencrypt.log`?

bradmw replied:
Not sure what you mean here?

kuba replied:
Sometimes you would like to debug what exactly happened, but you would need to mess with the source code and comment out some stuff in order to do so. Since we need logging to be set up in order to write full traceback to `--logs-dir`, it's not an option to use it. But you can attempt to write the full traceback to the current directory, mimicking core dump files.

kuba replied:
also, I think it'd be good to `assert exc_value` to make sure than non-zero exit code is returned

bradmw replied:
Ah gotcha. That sounds like a good idea to me. I'll implement it tomorrow. Thanks for the review @kuba.

kuba commented on letsencrypt/tests/cli_test.py:
```diff
@@ -66,12 +66,30 @@ def test_exceptions(self):
+        # pylint: disable=protected-access
+        import StringIO
+        from letsencrypt import cli
+        from letsencrypt import errors
```
modules that are not under test (here: `StringIO`, `errors`) should be imported at the top of the file

kuba commented on letsencrypt/tests/cli_test.py:
```diff
@@ -66,12 +66,30 @@ def test_exceptions(self):
+        from letsencrypt import cli
+        from letsencrypt import errors
+
+        cli._handle_exception(errors.Error, "detail", None, None)
```
`(errors.Error, exc_value="detail", trace=None, args=None)` could make it a bit more readable without having to jump around the files, same below

kuba commented on letsencrypt/cli.py:
```diff
@@ -642,8 +644,45 @@ def _setup_logging(args):
+            "An unexpected error occurred. Please see the logfiles in {0} for "
+            "more details.".format(args.logs_dir))
+    else:
+        traceback.print_exception(exc_type, exc_value, trace, file=sys.stderr)
```
does this exit with non-zero error code?

kuba replied:
and if so, then this should be documented inline

kuba commented:
Nice work, @bradmw. I like this approach a lot more! :)

Could you please make sure to mock out `_handle_exception` in all other unit tests in `cli_test.py`? This is regression introduced in #562 and reinforced here: exceptions will be hidden during tests. Note, that some unit tests don't even check the returned values but expect the exception to be raised instead, e.g. `test_plugins` (https://github.com/letsencrypt/letsencrypt/blob/abe1aa999af7b45e28958a7dcdd8fcc0daa310dd/letsencrypt/tests/cli_test.py#L60).

kuba commented:
Ah, and integration tests should include `--debug` IMHO.

bradmw commented:
Cleaned up the code and added support for a basic logfile in the working directory. I think this slightly overkill from what I previously had, but it can't hurt.

@kuba, thanks again for the feedback. The only changes you suggested and I didn't make are:
1. I don't `assert exc_value` because `sys.exit(exc_value)` is only called when `exc_type` is a subclass of `errors.Error`. This means that `exc_value` should be an instance of one of our exception types and `sys.exit` will always return a nonzero value.
2. I didn't mock out `_handle_exception` in other unit tests. I did some testing and it appears that unittest catches and reports the exception before `sys.excepthook` is called.

kuba commented on letsencrypt/cli.py:
```diff
@@ -642,8 +644,70 @@ def _setup_logging(args):
+        elif args is None:
+            sys.exit(
+                "An unexpected error occurred. Please see the logfile '{0}' "
+                "for more details.".format(os.path.abspath("letsencrypt.log")))
```
I'd personally remove `abspath`. It makes it more clear that that dump is in the current working directory.

kuba commented:
1. Yup, that sounds reasonable.
2. Having checked that, `unittest` seems to do some black voodo magic to custom `sys.excepthook` - namely ignore it...

LGTM :+1:. Feel free to merge, but consider adjusting `abspath` before or in a follow-up PR.


BTW, we also need a follow-up PR to fix the following:
```
    raise messages.Error.from_json(jobj)
Error: serverInternal :: The server experienced an internal error :: Error creating new authz
```
gets turned into `An unexpected error occurred. Please see the logfiles in /tmp/le/logs for more details.` when run without `--debug`, but I guess we should explicitly catch `messages.Error` in `main`.

jdkasten commented:
Thanks!
